src/FSharp.Core/local.fs

```diff
@@ -480,17 +480,57 @@ module internal List =
             else
                 filterToFreshConsTail cons f t
 
-    let rec filter predicate l =
-        match l with
-        | [] -> l
-        | h :: ([] as nil) -> if predicate h then l else nil
-        | h :: t ->
-            if predicate h then
-                let cons = freshConsNoTail h
-                filterToFreshConsTail cons predicate t
-                cons
-            else
-                filter predicate t
+    [<System.Runtime.CompilerServices.MethodImpl(System.Runtime.CompilerServices.MethodImplOptions.NoInlining)>]
+    let invalidStateOfEmptyList () = raise (System.Exception "logic error; list should not be empty")
+
+    /// optimized filter if filtered items are all at the end of the list then that list is returned
+    /// the function entryToFilterToFreshConsTail was the original version of filter, and is used
+    /// in the case of there being multiple segments in the data
+    let filter predicate originalList =
+        let rec entryToFilterToFreshConsTail lst =
+            match lst with
+            | [] -> []
+            | hd :: tl when predicate hd ->
+                match tl with 
+                | [] -> lst
+                | _ -> 
+                    let cons = freshConsNoTail hd
+                    filterToFreshConsTail cons predicate tl
+                    cons
+            | _ :: tl -> entryToFilterToFreshConsTail tl
+
+        let rec prependFirstSegmentItems exclusiveFirstSegmentEnd firstSegmentPtr newListWithNoTail =
+            match firstSegmentPtr with
+            | [] -> invalidStateOfEmptyList ()
+            | _ when obj.ReferenceEquals (firstSegmentPtr, exclusiveFirstSegmentEnd) -> newListWithNoTail
```

Comment:
We've recently discussed usage of `ReferenceEquals` for specific optimizaitons, its relations to structural equality and how it can change semantics of the code and differences in Release/Debug.

thoughts @T-Gro @dsyme?

Reply:
IMO here it is used in a controlled scenario, so at least the structural equality impact should not apply.
(because even if the same list reference with e.g. `[5.0;nan;0.0]' is returned after filtering, it won't be equal to itself)